src/garage/torch/policies/context_conditioned_policy.py

```diff
@@ -0,0 +1,268 @@
+        self._context_encoder.reset(num_tasks)
+
+    def sample_from_belief(self):
+        """Sample z using distributions from current means and variances."""
+        if self._use_ib:
+            posteriors = [
+                torch.distributions.Normal(m, torch.sqrt(s)) for m, s in zip(
+                    torch.unbind(self.z_means), torch.unbind(self.z_vars))
+            ]
+            z = [d.rsample() for d in posteriors]
+            self.z = torch.stack(z)
+        else:
+            self.z = self.z_means
+
+    def detach_z(self):
+        """Disable backprop through z."""
+        self.z = self.z.detach()
+
+    def update_context(self, inputs):
+        """Append single transition to the current context.
+
+        Args:
+            inputs (dict): Dictionary of transition information in np arrays .
+
+        """
+        o, a, r, no, _, _ = inputs
+        o = torch.as_tensor(o[None, None, ...],
+                            device=tu.global_device()).float()
+        a = torch.as_tensor(a[None, None, ...],
+                            device=tu.global_device()).float()
+        r = torch.as_tensor(np.array([r])[None, None, ...],
+                            device=tu.global_device()).float()
+        no = torch.as_tensor(no[None, None, ...],
+                             device=tu.global_device()).float()
+
+        if self._use_next_obs:
+            data = torch.cat([o, a, r, no], dim=2)
+        else:
+            data = torch.cat([o, a, r], dim=2)
+
+        if self._context is None:
+            self._context = data
+        else:
+            self._context = torch.cat([self._context, data], dim=1)
+
+    def infer_posterior(self, context):
+        """Compute q(z|c) as a function of input context and sample new z.
+
+        Args:
+            context (torch.Tensor): Context values.
+
+        """
+        params = self._context_encoder(context)
+        params = params.view(context.size(0), -1,
+                             self._context_encoder._output_dim)
+        # with probabilistic z, predict mean and variance of q(z | c)
+        if self._use_ib:
+            mu = params[..., :self._latent_dim]
+            sigma_squared = F.softplus(params[..., self._latent_dim:])
+            z_params = [
+                tu.product_of_gaussians(m, s)
+                for m, s in zip(torch.unbind(mu), torch.unbind(sigma_squared))
+            ]
+            self.z_means = torch.stack([p[0] for p in z_params])
+            self.z_vars = torch.stack([p[1] for p in z_params])
+        else:
+            self.z_means = torch.mean(params, dim=1)
+        self.sample_from_belief()
+
+    # pylint: disable=arguments-differ
+    def forward(self, obs, context):
+        """Given observations and context, get actions and probs from policy.
+
+        Args:
+            obs (torch.Tensor): Observation values.
+            context (torch.Tensor): Context values.
+
+        Returns:
+            tuple:
```

Comment:
which of these is used most commonly?

Reply:
basically, i'm asking if this is really the best return signature.

if most people just want `z`, then i suggest your signature look like

```python
"""
    Returns:
        torch.Tensor: z values....
        tuple: A bunch of the other stuff
"""

# This would give you a callsite like this
z, _ = Encoder.forward(stuff)
```
This pattern puts the most common output as the first, and then if you really need something you can unpack what we throw away in `_`.

Another question I have is, "Do you have examples of your code using every element of the return tuple?" If not, I suggest you just remove them. Big APIs are hard to keep promises for. If you use one part of a symmetric pair, such as mean but not log_std, it's fine to also return log_std. But I'm skeptical that all of these values are used.